packages/eslint-mdx/src/tokens.ts

```diff
@@ -0,0 +1,257 @@
+      // const attrEnd = attrPos.end.offset
+
+      const attrStart = nextCharOffset(lastAttrOffset + 1)
+
+      assert(attrStart != null)
+
+      const attrName = attr.name
+      const attrNameLength = attrName.length
+
+      tokens.push(
+        newToken(tt.jsxName, attrStart, attrStart + attrNameLength, attrName),
+      )
+
+      const attrValue = attr.value
+
+      if (!attrValue) {
+        lastAttrOffset = attrStart + attrNameLength
+        continue
+      }
+
+      const attrEqualOffset = nextCharOffset(attrStart + attrNameLength)
+
+      assert(text[attrEqualOffset] === '=')
+
+      tokens.push(newToken(tt.eq, attrEqualOffset, attrEqualOffset + 1, '='))
+
+      // `mdxJsxAttributeValueExpression`, already handled by acorn
+      if (typeof attrValue === 'object') {
+        const data = attrValue.data
+
+        let [attrValStart, attrValEnd] = data.estree.range
+
+        attrValStart = prevCharOffset(attrValStart - 1)
+        attrValEnd = nextCharOffset(attrValEnd)
+
+        assert(text[attrValStart] === '{')
+        assert(text[attrValEnd] === '}')
+
+        lastAttrOffset = attrValEnd
+
+        tokens.push(
+          newToken(tt.braceL, attrValStart, attrValStart + 1),
+          newToken(tt.braceR, attrValEnd, attrValEnd + 1),
+        )
+
+        continue
+      }
+
+      const attrQuoteOffset = nextCharOffset(attrEqualOffset + 1)
+
+      const attrQuote = text[attrQuoteOffset]
+
+      assert(attrQuote === '"' || attrQuote === "'")
+
+      tokens.push(
+        newToken(
+          tt.string,
+          attrQuoteOffset,
+          attrQuoteOffset + attrValue.length + 2,
+          attrValue,
+        ),
+      )
+
+      lastAttrOffset = nextCharOffset(attrQuoteOffset + attrValue.length + 1)
+
+      assert(text[lastAttrOffset] === attrQuote)
+    }
+
+    const nextOffset = nextCharOffset(lastAttrOffset + 1)
+
+    const nextChar = text[nextOffset]
+
+    // self closing tag
+    if (nextChar === '/') {
+      tokens.push(newToken(tt.slash, nextOffset, nextOffset + 1, '/'))
+    } else {
+      assert(nextChar === '>')
+
+      const prevOffset = prevCharOffset(nodeEnd - 2)
+
+      if (nodeName) {
+        tokens.push(
+          newToken(
+            tt.jsxName,
+            prevOffset + 1 - nodeNameLength,
+            prevOffset + 1,
+            nodeName,
+          ),
+        )
+      }
+
+      const slashOffset = prevCharOffset(prevOffset - nodeNameLength)
+
+      assert(text[slashOffset] === '/')
+
+      tokens.push(newToken(tt.slash, slashOffset, slashOffset + 1, '/'))
+
+      const tagStartOffset = prevCharOffset(slashOffset - 1)
+
+      assert(text[tagStartOffset] === '<')
+
+      tokens.push(newToken(tt.jsxTagStart, tagStartOffset, tagStartOffset + 1))
+    }
+
+    tokens.push(newToken(tt.jsxTagEnd, nodeEnd - 1, nodeEnd))
+  })
+
+  tokens.push(newToken(tt.eof, text.length, text.length))
+
+  return tokens
+    .filter(
+      t =>
+        !(
+          t.start === t.end &&
```

Comment:
I should be added incorrectly by `remark-mdx` somewhere. @wooorm

Reply:
I don’t quite understand what you mean?

Do you mean that, tokens generated by Acorn for `prefix` and `suffix` should *not* be passed to [`onToken`](https://github.com/micromark/micromark-extension-mdx-expression/blob/efdcf84d634cff622b20c30dd9249f00da6f3554/packages/micromark-util-events-to-acorn/dev/index.js#L45)?

Reply:
Yep, I was surprised by these tokens when I tested locally. 😂

Reply:
ahh, that makes sense. Is this required for your code to work? Can you open an issue in `micromark-extension-mdx-expression` about `prefix`/`suffix` tokens being passed to `onToken`?

Reply:
Done. https://github.com/micromark/micromark-extension-mdx-expression/issues/6